File: backend/app/utils/formatters.py
from datetime import datetime, date, timedelta


def format_date(date_obj, format_type='long'):
    d = datetime.fromisoformat(date_obj) if isinstance(date_obj, str) else date_obj
    if format_type == 'short':
        return d.strftime("%d %b")
    if format_type == 'long':
        return d.strftime("%d %B %Y")
    if format_type == 'relative':
        now = datetime.now()
        diff_days = (now - d).days
        if diff_days == 0:
            return "Today"
        if diff_days == 1:
            return "Yesterday"
        if diff_days < 7:
            return f"{diff_days} days ago"
        return d.strftime("%d %b")
    return d.strftime("%Y-%m-%d")


def format_time(date_obj):
    d = datetime.fromisoformat(date_obj) if isinstance(date_obj, str) else date_obj
    return d.strftime("%I:%M %p").lstrip("0")


def format_score(score):
    return f"{min(100, max(0, score)):.0f}%"


def get_score_color(score):
    if score >= 80:
        return "text-green-500"
    if score >= 60:
        return "text-yellow-500"
    if score >= 40:
        return "text-orange-500"
    return "text-red-500"


def truncate_text(text, max_length=100):
    if len(text) <= max_length:
        return text
    return text[:max_length] + "..."


def generate_certificate_id():
    import random
    import string
    return "VANI-" + ''.join(random.choices(string.ascii_uppercase + string.digits, k=10))

File: frontend/src/utils/formatters.js
export const formatDate = (date, format = 'long') => {
  const d = new Date(date)
  if (format === 'short') {
    return d.toLocaleDateString('en-IN', { day: 'numeric', month: 'short' })
  }
  if (format === 'long') {
    return d.toLocaleDateString('en-IN', { day: 'numeric', month: 'long', year: 'numeric' })
  }
  if (format === 'relative') {
    const now = new Date()
    const diffDays = Math.floor((now - d) / (1000 * 60 * 60 * 24))
    if (diffDays === 0) return 'Today'
    if (diffDays === 1) return 'Yesterday'
    if (diffDays < 7) return `${diffDays} days ago`
    return formatDate(date, 'short')
  }
  return d.toLocaleDateString()
}

export const formatTime = (date) => {
  return new Date(date).toLocaleTimeString('en-IN', { hour: '2-digit', minute: '2-digit' })
}

export const formatScore = (score) => {
  const num = Math.min(100, Math.max(0, score))
  return `${Math.round(num)}%`
}

export const getScoreColor = (score) => {
  if (score >= 80) return 'text-green-500'
  if (score >= 60) return 'text-yellow-500'
  if (score >= 40) return 'text-orange-500'
  return 'text-red-500'
}

export const truncateText = (text, maxLength = 100) => {
  if (text.length <= maxLength) return text
  return text.slice(0, maxLength) + '...'
}

export const generateCertificateId = () => {
  return 'VANI-' + Math.random().toString(36).substring(2, 10).toUpperCase()
}